### ISSUE TITLE
server: SPDY error when accessing admin UI with chrome 

### ISSUE THREAD

**Issue by mberhault** [cockroachdb/cockroach#5020]
between a3cff59889f3c6094a5b5ed310afd8eab5fde0d5 and 9136aef1d05aab2385c2c5a2402c0a95f793cbb3, hitting the admin page in secure mode fails in chrome with `ERR_SPDY_PROTOCOL_ERROR`.

I'll see if I can grab some network traffic. So far, nothing shows up in the debug window.

**Comment by tamird** [comment]
What do you mean "between"?

**Comment by petermattis** [comment]
Didn't Chrome remove support for `SPDY`? Perhaps you need to upgrade Chrome.

**Comment by mberhault** [comment]
I mean between those two shas. Specifically, I'm guessing https://github.com/cockroachdb/cockroach/commit/3192a60948d56e5bb79e0fbbc6c27a6cbb007237 is the culprit.

**Comment by mberhault** [comment]
I'm on  48.0.2564.116, but I'll restart, there is a new stable release of chrome. brb

**Comment by mberhault** [comment]
No luck with latest stable: 49.0.2623.75

**Comment by tamird** [comment]
Seems likely.

Does this work correctly in insecure mode?

**Comment by mberhault** [comment]
last event in the chrome net-trace:
```
5133: HTTP2_SESSION
localhost:26257 (DIRECT)
Start Time: 2016-03-09 13:18:20.509

t=1678 [st=0] +HTTP2_SESSION  [dt=1]
               --> host = "localhost:26257"
               --> proxy = "DIRECT"
t=1678 [st=0]    HTTP2_SESSION_INITIALIZED
                 --> protocol = "h2"
                 --> source_dependency = 5132 (SOCKET)
t=1678 [st=0]    HTTP2_SESSION_SEND_SETTINGS
                 --> settings = ["[id:3 flags:0 value:1000]","[id:4 flags:0 value:6291456]"]
t=1678 [st=0]    HTTP2_STREAM_UPDATE_RECV_WINDOW
                 --> delta = 15663105
                 --> window_size = 15728640
t=1678 [st=0]    HTTP2_SESSION_SENT_WINDOW_UPDATE_FRAME
                 --> delta = 15663105
                 --> stream_id = 0
t=1678 [st=0]    HTTP2_SESSION_SEND_HEADERS
                 --> fin = true
                 --> :method: GET
                     :authority: localhost:26257
                     :scheme: https
                     :path: /
                     pragma: no-cache
                     cache-control: no-cache
                     accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
                     upgrade-insecure-requests: 1
                     user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.75 Safari/537.36
                     accept-encoding: gzip, deflate, sdch
                     accept-language: en-US,en;q=0.8,fr;q=0.6
                 --> priority = 0
                 --> stream_id = 1
                 --> unidirectional = false
t=1679 [st=1]    HTTP2_SESSION_CLOSE
                 --> description = "Framer error: 1 (INVALID_CONTROL_FRAME)."
                 --> net_error = -337 (ERR_SPDY_PROTOCOL_ERROR)
t=1679 [st=1]    HTTP2_SESSION_POOL_REMOVE_SESSION
                 --> source_dependency = 5133 (HTTP2_SESSION)
t=1679 [st=1] -HTTP2_SESSION
```

**Comment by mberhault** [comment]
and yes, insecure mode is happy.

**Comment by tamird** [comment]
OK, I think I know what's going on here.

**Comment by tamird** [comment]
Another cmux bug.

**Comment by mberhault** [comment]
interestingly, when listing HTTP2 sessions through chrome://net-internals, I see the failed ones (as listed above) in secure mode, but nothing shows up in insecure mode.

**Comment by tamird** [comment]
I'm guessing http1 is being used in secure mode - chrome doesn't do h2c.

**Comment by tamird** [comment]
TL;DR: things are broken. Most http2 clients (chrome and go both included) behave as follows (after the preface):
- send settings
- send window update
- wait for settings ack
- send headers

grpc behaves differently: it doesn't wait for the settings ack. This means we can match grpc connections, but non-grpc H2 connections are **broken** - they will hang indefinitely while we wait for them to send us headers so we can figure out if they're grpc connections or not.

I'm still trying to work out a solution.
